`SyntaxError: Unexpected token M in JSON at position 82780` when doing a `terraform show ./tfplan`

**Describe the bug**

Azure DevOps is reporting an error when the pipeline runs a `terraform show -json ./tfplan` task. The error says:

```
2021-07-01T06:22:45.4636143Z ##[error]SyntaxError: Unexpected token M in JSON at position 82780
```

Full debug output from the task is available at https://gist.github.com/ausfestivus/9eca2d6d03aa81ff7a2c3251335b5eb9

**To Reproduce**

The error occurs with this terraform code. When I run the same pipeline YAML with a different set of terraform files the problem does not occur.

**Expected behavior**

The error should not occur and the terraform command should succeed.

**Screenshots**

n/a

**Pipeline Logs**

See gist https://gist.github.com/ausfestivus/9eca2d6d03aa81ff7a2c3251335b5eb9

**Agent Configuration**

 - OS: ubuntu 20 LTS
 - Self hosted
 - 0.13.7
 - AzureCLI version used (Default for hosted agent is acceptable or N/A)

**Additional context**

I suspect this is an AzDO problem but I am stuck on how to troubleshoot further. A google search turned up a link to this repo at https://github.com/charleszipp/azure-pipelines-tasks-terraform/issues/171 but that URL 404s. Its probably an issue in the old repo.

**EDIT**

* corrected the command that is being run that produces the error

This issue has just turned up again in my pipeline.

```
2021-11-08T00:53:04.9955647Z ##[error]SyntaxError: Unexpected token M in JSON at position 188984
2021-11-08T00:53:04.9958173Z ##[debug]Processed: ##vso[task.issue type=error;]SyntaxError: Unexpected token M in JSON at position 188984
```

```
2021-11-08T00:53:02.8627026Z ##[debug]/opt/hostedtoolcache/terraform/0.15.5/x64/terraform arg: show
2021-11-08T00:53:02.8627977Z ##[debug]/opt/hostedtoolcache/terraform/0.15.5/x64/terraform arg: -json
2021-11-08T00:53:02.8628840Z ##[debug]/opt/hostedtoolcache/terraform/0.15.5/x64/terraform arg: ./tfplan
```

```
2021-11-08T00:53:02.3885672Z ##[section]Starting: terraform show plan
2021-11-08T00:53:02.3893932Z ==============================================================================
2021-11-08T00:53:02.3894218Z Task         : Terraform CLI
2021-11-08T00:53:02.3894455Z Description  : Execute terraform cli commands
2021-11-08T00:53:02.3894653Z Version      : 0.6.28
2021-11-08T00:53:02.3894852Z Author       : Charles Zipp
2021-11-08T00:53:02.3895041Z Help         : 
2021-11-08T00:53:02.3895274Z ==============================================================================
```

--- a/tasks/terraform-cli/src/commands/tf-show.ts
+++ b/tasks/terraform-cli/src/commands/tf-show.ts
@@ -8,15 +8,15 @@ import { ITaskAgent } from "../task-agent";
 export class TerraformShow implements ICommand {
     constructor(
         private readonly taskAgent: ITaskAgent,
-        private readonly runner: IRunner, 
+        private readonly runner: IRunner,
         private readonly logger: ILogger
         ) {
     }
 
-    async exec(ctx: ITaskContext): Promise<CommandResponse> {        
+    async exec(ctx: ITaskContext): Promise<CommandResponse> {
         const options = await new RunWithTerraform(ctx, true)
-            .withSecureVarFile(this.taskAgent, ctx.secureVarsFileId, ctx.secureVarsFileName)    
-            .withJsonOutput(ctx.commandOptions)            
+            .withSecureVarFile(this.taskAgent, ctx.secureVarsFileId, ctx.secureVarsFileName)
+            .withJsonOutput(ctx.commandOptions)
             .withCommandOptions(ctx.commandOptions)
             .withPlanOrStateFile(ctx.planOrStateFilePath)
             .build();
@@ -40,7 +40,8 @@ export class TerraformShow implements ICommand {
 
     private detectDestroyChanges(ctx: ITaskContext, result: string): void
     {
-        const resultNoEol = result.replace(/(\r\n|\r|\n|\\n|\t|\\")/gm, "");
+        // Remove unexpected characters at the end of the result string
+        const resultNoEol = result.replace(/(\r\n|\r|\n|\\n|\t|\\")$/gm, "");
         let jsonResult = JSON.parse(resultNoEol);
         const deleteValue = "delete";
 
@@ -54,7 +55,7 @@ export class TerraformShow implements ICommand {
                 }
             }
         }
-        
+
         this.logger.debug("No destroy detected")
         this.setDestroyDetectedFlag(ctx, false);
     }

--- a/tasks/terraform-cli/src/tests/features/terraform-show.feature
+++ b/tasks/terraform-cli/src/tests/features/terraform-show.feature
@@ -33,7 +33,7 @@ Feature: terraform show
         And the terraform cli task is successful
         And pipeline variable "TERRAFORM_LAST_EXITCODE" is set to "0"
         And pipeline variable "TERRAFORM_PLAN_HAS_DESTROY_CHANGES" is set to "true"
-        
+
         Scenario: show tf plan file with destroy operations - apiManagement
         Given terraform exists
         And terraform command is "show"
@@ -97,6 +97,17 @@ Feature: terraform show
         Given terraform exists
         And terraform command is "show"
         And running command "terraform show -json show.plan" returns successful result with stdout from file "./src/tests/stdout_tf_show_tfplan_output_only.json"
+        And the target plan or state file is "show.plan"
+        When the terraform cli task is run
+        Then the terraform cli task executed command "terraform show -json show.plan"
+        And the terraform cli task is successful
+        And pipeline variable "TERRAFORM_LAST_EXITCODE" is set to "0"
+        And pipeline variable "TERRAFORM_PLAN_HAS_DESTROY_CHANGES" is set to "false"
+
+        Scenario: show tf plan that contains newlines
+        Given terraform exists
+        And terraform command is "show"
+        And running command "terraform show -json show.plan" returns successful result with stdout from file "./src/tests/stdout_tf_show_tfplan_with_newline.json"
         And the target plan or state file is "show.plan"
         When the terraform cli task is run
         Then the terraform cli task executed command "terraform show -json show.plan"

--- a/tasks/terraform-cli/src/tests/stdout_tf_show_tfplan_with_newline.json
+++ b/tasks/terraform-cli/src/tests/stdout_tf_show_tfplan_with_newline.json
@@ -1,0 +1,23 @@
+{
+    "format_version": "0.1",
+    "terraform_version": "0.13.4",
+    "planned_values": {
+        "outputs": {
+            "some_string": {
+                "sensitive": false,
+                "value": "value\\nwith\\nnewline\\n"
+            }
+        },
+        "root_module": {}
+    },
+    "output_changes": {
+        "some_string": {
+            "actions": [
+                "create"
+            ],
+            "before": null,
+            "after": "value\\nwith\\nnewline\\n",
+            "after_unknown": false
+        }
+    }
+}

